### PR TITLE
New module: ec2_instance

### PR DESCRIPTION
**Pull request by ryansb**
##### ISSUE TYPE

 - Feature Pull Request
 - New Module Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### ANSIBLE VERSION

```

```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 1]
```diff
@@ -0,0 +1,1300 @@
...
+      - i-12345678
+
+# Gather facts about any instance with a tag key Name and value Example
+- ec2_instance_facts:
+    filters:
+      "tag:Name": Example
+'''
+
+RETURN = '''
+instances:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py:185]: spec is also returned

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 2]
```diff
@@ -0,0 +1,1300 @@
...
+                                      ec2_argument_spec,
+                                      get_aws_connection_info,
+                                      AWSRetry,
+                                      ansible_dict_to_boto3_filter_list,
+                                      compare_aws_tags,
+                                      boto3_tag_list_to_ansible_dict,
+                                      ansible_dict_to_boto3_tag_list,
+                                      camel_dict_to_snake_dict)
+
+from ansible.module_utils.aws.core import AnsibleAWSModule, HAS_BOTO3
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py]: The import for HAS_BOTO3 can be removed

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 3]
```diff
@@ -0,0 +1,1300 @@
...
+
+    interfaces = []
+    network = params.get('network') or {}
+    if network or not params.get('network_interfaces'):
+        # they only specified one interface
+        spec = {
+            'DeviceIndex': 0,
+        }
+        if network.get('source_dest_check') is not None:
+            raise Exception("Uhoh - srcdest not supported")
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py]: You can use module.fail_json() instead

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 4]
```diff
@@ -0,0 +1,1300 @@
...
+        spec['ImageId'] = params['image_id']
+    elif isinstance(params.get('image'), dict):
+        image = params.get('image', {})
+        spec['ImageId'] = image.get('id')
+        if 'ramdisk' in image:
+            spec['RamdiskId'] = image['ramdisk']
+        if 'kernel' in image:
+            spec['KernelId'] = image['kernel']
+    if not spec.get('ImageId') and not params.get('launch_template'):
+        raise ValueError("You must include an image_id or image.id parameter to create an instance")
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py]: This isn't caught anywhere so it gives an unformatted traceback. fail_json() could be used.

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 5]
```diff
@@ -0,0 +1,1300 @@
...
+            {
+                'Primary': True|False,
+                'PrivateIpAddress': 'string'
+            },
+        ],
+        'SecondaryPrivateIpAddressCount': 123,
+        'SubnetId': 'string'
+    },
+    """
+    if ec2 is None:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py:725]: This check is done in a number of places. Is there a reason to not instantiate it once in main()?

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 6]
```diff
@@ -0,0 +1,1300 @@
...
+        # TODO use default filters to find instances
+        # probably subnet-id + name
+        return []
+        module.fail_json(msg="No filters provided when they were required")
+        base_filters = {
+            'subnet-id': module.params.get('name')
+        }
+        if module.params.get('vpc_subnet_id'):
+            base_filters['subnet-id'] = module.params.get('vpc_subnet_id')
+        if module.params.get('name'):
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py]: name isn't in the docs yet. Also, should there be a check for the tag 'Name'?

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 7]
```diff
@@ -0,0 +1,1300 @@
...
+        Filters=ansible_dict_to_boto3_filter_list({
+            'vpc-id': vpc['VpcId'],
+            'state': 'available',
+            'default-for-az': 'true',
+        })
+    )
+    if len(subnets.get('Subnets', [])):
+        # to have a deterministic sorting order, we sort by AZ so we'll always pick the `a` subnet first
+        # there can only be one default-for-az subnet per AZ, so the AZ key is always unique in this list
+        by_az = sorted(subnets.get('Subnets'), key=lambda s: s['AvailabilityZone'])
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py]: Nice :)

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 8]
```diff
@@ -0,0 +1,1300 @@
...
+        iam = module.client('iam')
+    try:
+        role = iam.get_instance_profile(InstanceProfileName=name_or_arn)
+        return role['InstanceProfile']['Arn']
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'NoSuchEntity':
+            module.fail_json_aws(msg="Could not find instance_role {0}".format(name_or_arn))
+        module.fail_json_aws(msg="An error occurred while searching for instance_role "
+                                 "{0}. Please try supplying the full ARN.".format(name_or_arn))
+    module.fail_json(msg="Could not retrieve instance_role {0}".format(name_or_arn))
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py]: I think if `return role['InstanceProfile']['Arn']` fails, that should raise an Exception, right? I don't think this line gets used.

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 9]
```diff
@@ -0,0 +1,1300 @@
...
+    state = module.params.get('state')
+    ec2 = module.client('ec2')
+    if module.params.get('filters') is None:
+        filters = {
+            # all states except shutting-down and terminated
+            'instance-state-name': ['pending', 'running', 'stopping', 'stopped']
+        }
+        if state == 'stopped':
+            # only need to change instances that aren't already stopped
+            filters['instance-state-name'] = ['pending', 'running']
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py]: If they want to wait for an instance that is 'stopping' but not 'stopped' should that be possible?

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 10]
```diff
@@ -0,0 +1,1300 @@
...
+            if state in ('running', 'started') and [i for i in existing_matches if i['State']['Name'] != 'running']:
+                ins_changed, failed, instances = change_instance_state(filters=module.params.get('filters'), desired_state='RUNNING')
+                module.exit_json(
+                    changed=bool(len(ins_changed)) or changed,
+                    instances=[pretty_instance(i) for i in instances],
+                    instance_ids=[i['InstanceId'] for i in instances],
+                )
+            changes = diff_instance_and_params(existing_matches[0], module.params)
+            for c in changes:
+                module.warn(str(c))
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py]: Why is it warning here?

Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_instance.py]: That was debugging. Removed.

File: lib/ansible/modules/cloud/amazon/ec2_instance.py (hunk 9 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_instance.py]: Yeah, we can make that a thing. Implemented.

File: lib/ansible/modules/cloud/amazon/ec2_instance.py (hunk 8 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_instance.py]: Correct. Removed.

File: lib/ansible/modules/cloud/amazon/ec2_instance.py (hunk 4 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_instance.py]: Good catch

File: lib/ansible/modules/cloud/amazon/ec2_instance.py (hunk 2 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_instance.py]: Done.

File: lib/ansible/modules/cloud/amazon/ec2_instance.py (hunk 1 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_instance.py:185]: I'm torn on whether I should return `spec` - do you think it's useful to leave as a debugging tool to show people what instance we ran?

File: lib/ansible/modules/cloud/amazon/ec2_instance.py (hunk 3 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_instance.py]: This is removed - source_dest_check is now just plain supported.

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
Temporarily commenting out unrelated test failures in https://github.com/ansible/ansible/pull/35883.

**Comment by s-hertel** [comment]
rebuild_merge

**Comment by s-hertel** [comment]
Transient failures are unrelated.

**Pull request merged**